pkg/experiment/metastore/metastore_readindex.go

```diff
@@ -5,128 +5,65 @@ import (
-				return &metastorev1.ReadIndexResponse{ReadIndex: readIndex}, nil
+			if appliedIndex >= r.ReadIndex {
+				return nil
 			}
-			continue
-		case <-timeout.C:
-			raftLogger().Log("err", "timeout")
-			return new(metastorev1.ReadIndexResponse), fmt.Errorf("timeout")
 		case <-ctx.Done():
-			raftLogger().Log("err", "context canceled")
-			return new(metastorev1.ReadIndexResponse), fmt.Errorf("canceled %w", ctx.Err())
+			return ctx.Err()
 		}
 	}
 }
 
-func (m *Metastore) CheckReady(ctx context.Context) (err error) {
-	const (
-		ready    = "ready"
-		notReady = "not_ready"
-		status   = "status"
-	)
-	debugRequestId := uuid.Must(uuid.NewRandom()).String() //todo delete
-	readIndex := uint64(0)
-	t := time.Now()
-	raftLogger := func() log.Logger {
-		return log.With(m.logger, "component", "raft_debug",
-			"request_id", debugRequestId,
-			"op", "CheckReady",
-			"read_index", readIndex,
-			"applied_index", m.raft.AppliedIndex(),
-			"commit_index", m.raft.CommitIndex(),
-			"last_index", m.raft.LastIndex(),
-			"duration", time.Since(t),
-		)
-	}
-	raftLogger().Log("msg", "check")
-	req := new(metastorev1.ReadIndexRequest)
-	req.DebugRequestId = debugRequestId
-	res, err := m.client.ReadIndex(ctx, req)
-	if err != nil {
-		err = fmt.Errorf("failed to get read index: %w", err)
-		raftLogger().Log(status, notReady, "err", err)
+// CheckReady verifies if the node is ready to serve requests.
+// It ensures the node is up-to-date by:
+// 1. Performing a ReadIndex operation
+// 2. Waiting for the applied index to catch up to the read index
+// 3. Enforcing a minimum duration since first becoming ready
+func (m *Metastore) CheckReady(ctx context.Context) error {
+	if err := m.readIndex(ctx); err != nil {
 		return err
 	}
-	readIndex = res.ReadIndex
-
-	tcheck := time.NewTicker(tcheckFreq)
-	defer tcheck.Stop()
-	timeout := time.NewTimer(5 * time.Second)
-	defer timeout.Stop()
 
-	for {
-		select {
-		case <-tcheck.C:
-			commitIndex := m.raft.CommitIndex()
-			raftLogger().Log("msg", "tick")
-			if commitIndex >= res.ReadIndex {
-				if m.readySince.IsZero() {
-					m.readySince = time.Now()
-				}
-				minReadyTime := m.config.MinReadyDuration
-				if time.Since(m.readySince) < minReadyTime {
-					err := fmt.Errorf("waiting for %v after being ready", minReadyTime)
-					raftLogger().Log(status, notReady, "err", err)
-					return err
-				}
+	if m.readySince.IsZero() {
+		m.readySince = time.Now()
+	}
 
-				raftLogger().Log(status, ready)
-				return nil
-			}
-			continue
-		case <-timeout.C:
-			raftLogger().Log(status, notReady, "err", "timeout")
-			return fmt.Errorf("metastore ready check timeout")
-		case <-ctx.Done():
-			raftLogger().Log(status, notReady, "err", "context canceled")
-			return fmt.Errorf("metastore check context canceled %w", ctx.Err())
-		}
+	minReadyTime := m.config.MinReadyDuration
+	if time.Since(m.readySince) < minReadyTime {
+		return fmt.Errorf("waiting for %v after being ready", minReadyTime)
 	}
+
+	return nil
```

Comment:
It does look like we're replicating k8s behaviour: https://kubernetes.io/blog/2021/08/27/minreadyseconds-statefulsets. I'm wondering if this is really necessary

Reply:
You're right.. so, I'd say we can remove it from the "generic" metastore config since we're duplicating behavioiur https://github.com/grafana/pyroscope/blob/main/pkg/experiment/metastore/metastore.go#L53